hypothesis-python/src/hypothesis/strategies/_internal/strings.py

```diff
@@ -159,6 +165,37 @@ def filter(self, condition):
                 f"all nonempty strings!  Did you mean str.is{condition.__name__}?",
                 HypothesisWarning,
             )
+        if condition is str.isidentifier:
+            from hypothesis.strategies._internal.core import from_regex
+
+            def _chars_to_regex_set(s):
+                spans = tuple((ord(c), ord(c)) for c in s)
+                return "".join(
+                    f"{chr(a)}" if a == b else f"{chr(a)}-{chr(b)}"
+                    for a, b in charmap._union_intervals(spans, spans)
+                )
```

Comment:
Looking at this with fresh eyes, the regex approach is quite inefficient - we would prefer to operate entirely in terms of spans (from our charmap helper module), rather than individual characters.

As an alternative, we could define `characters(...)` strategies directly for id_start and id_continue character sets [per this documentation](https://docs.python.org/3/reference/lexical_analysis.html#identifiers), and generate our identifiers from the intersection between these intervals and those of the filtered strategy.  

This should be much faster to compute, but we can still write exhaustive tests - which iterate over all unicode characters to confirm that we passed correct arguments to the id-strategies - and thus be confident enough to hard-code the various categories and special characters.

Reply:
I realize my new implementation is slower from combining two ``characters(...)`` strategies unlike using regex. However, I'm not sure how else to use ``characters(...)`` without doing this. I may be misinterpreting what you mean?

Reply:
Your intersection calculation is the better fast method I was talking about 😁 

If it's tricky to rebase on my recent regex-alphabet changes, let me know and I'm happy to do that bit.